src/lib/system/datasources/local_music_fetcher_impl.dart

```diff
@@ -30,200 +33,306 @@ class LocalMusicFetcherImpl implements LocalMusicFetcher {
+          return song.copyWith(color: color);
+        return song;
+      })
+      .toList();
+    }
+
+    asyncExecutor.close();
+
+
     return {
       'SONGS': songs,
       'ALBUMS': albums,
-      'ARTISTS': artistSet.toList(),
+      'ARTISTS': artists.toList(),
+    };
+  }
+
+  Future<List<File>> getSongFilesInDirectory(
+      String path,
+      Set<String> allowedExtensions,
+      Set<String> blockedPaths
+    ) async {
+    return Directory(path)
+        .list(recursive: true, followLinks: false)
+        .where((item) => FileSystemEntity.isFileSync(item.path))
+        .where((item) => !blockedPaths.contains(item.path))
+        .where((item) {
+          final extension = p.extension(item.path).toLowerCase().substring(1);
+          return allowedExtensions.contains(extension);
+        })
+        .asyncMap((item) => File(item.path))
+        .toList();
+  }
+
+  // Returns a list of all new song files and files that have changed since they where last imported
+  Future<List<File>> getSongFilesToCheck(List<File> songFiles) async {
+    final List<File> songFilesToCheck = [];
+
+    for (final songFile in songFiles) {
+      final lastModified = songFile.lastModifiedSync();
+      final song = await _musicDataSource.getSongByPath(songFile.path);
+
+      if (song == null || lastModified.isAfter(song.lastModified))
+        songFilesToCheck.add(songFile);
+    }
+
+    return songFilesToCheck;
+  }
+
+
+  Future<List<ArtistModel>> getSortedArtists() async {
+    return (await _musicDataSource.artistStream.first)
+      ..sort((a, b) => a.id.compareTo(b.id));
+  }
+
+  Future<List<AlbumModel>> getSortedAlbums() async {
+    return (await _musicDataSource.albumStream.first)
+      ..sort((a, b) => a.id.compareTo(b.id));
+  }
+
+  Future<int> getAlbumId(int newAlbumId, String? album, String? albumArtist, int? year) async {
+    return await _musicDataSource.getAlbumId(album, albumArtist, year) ?? newAlbumId++;
+  }
+
+  Future<String> cacheAlbumArt(Picture albumArt, int albumId) async {
+    final Directory dir = await getApplicationSupportDirectory();
+    final albumArtPath = '${dir.path}/$albumId';
+    final file = File(albumArtPath);
+    file.writeAsBytesSync(albumArt.data);
+
+    return albumArtPath;
+  }
+
+  // Maps all the songs that where scanned previously, and their Albums and Artists to the new data structures
+  Future<Map<String, dynamic>> mapSongsAlreadyScanned(
+    List<File> songFiles, 
+    List<AlbumModel> albumsInDb, 
+    List<ArtistModel> artistsInDb
+  ) async {
+    final List<SongModel> songs = [];
+    final List<AlbumModel> albums = [];
+    final Set<ArtistModel> artists = {};
+
+    final Map<String, int> albumIdMap = {};
+    final Map<int, String> albumArtMap = {};
+    /// album id, background color
+    final Map<int, Color?> colorMap = {};
+
+    for (final songFile in songFiles) {
+      final song = (await _musicDataSource.getSongByPath(songFile.path))!;
```

Comment:
In the branch for #124 I added a few performance improvements. Using a map instead of querying the db every time was one of them.

Reply:
If you want I can implement this optimization. This step is already the fastest though (taking up ~2 Seconds) so I don't think it is needed.

Reply:
Yeah, we can leave it for now. I optimized this to reduce the number of database queries, since they seemed to get slow for some users after a while. Let's just keep it in mind as an option.